kernel: Introduce a way to specify minimum system heap size

jhedberg opened the pull request:
There are several subsystems and boards which require a relatively large system heap (used by k_malloc()) to function properly. This became even more notable with the recent introduction of the ACPICA library, which causes ACPI-using boards to require a system heap of up to several megabytes in size.

Until now, subsystems and boards have tried to solve this by having Kconfig overlays which modify the default value of HEAP_MEM_POOL_SIZE. This works ok, except when applications start explicitly setting values in their prj.conf files:
```console
$ git grep CONFIG_HEAP_MEM_POOL_SIZE= tests samples|wc -l
     157
```
The vast majority of values set by current sample or test applications is much too small for subsystems like ACPI, which results in the application not being able to run on such boards.

To solve this situation, we introduce support for subsystems to specify their own custom system heap size requirement. Subsystems do this by defining Kconfig options with the prefix HEAP_MEM_POOL_ADD_SIZE_. The final value of the system heap is the sum of the custom minimum requirements, or the value existing HEAP_MEM_POOL_SIZE option, whichever is greater.

We also introduce a new HEAP_MEM_POOL_IGNORE_MIN Kconfig option which applications can use to force a lower value than what subsystems have specficied, however this behavior is disabled by default.

Whenever the minimum is greater than the requested value a CMake warning will be issued in the build output.

This patch ends up modifying several places outside of kernel code, since the presence of the system heap is no longer detected using a non-zero CONFIG_HEAP_MEM_POOL_SIZE value, rather it's now detected using a new K_HEAP_MEM_POOL_SIZE value that's evaluated at build time.

Fixes #65722

Commit:
kernel: Introduce a way to specify minimum system heap size

There are several subsystems and boards which require a relatively large
system heap (used by k_malloc()) to function properly. This became even
more notable with the recent introduction of the ACPICA library, which
causes ACPI-using boards to require a system heap of up to several
megabytes in size.

Until now, subsystems and boards have tried to solve this by having
Kconfig overlays which modify the default value of HEAP_MEM_POOL_SIZE.
This works ok, except when applications start explicitly setting values
in their prj.conf files:

$ git grep CONFIG_HEAP_MEM_POOL_SIZE= tests samples|wc -l
     157

The vast majority of values set by current sample or test applications
is much too small for subsystems like ACPI, which results in the
application not being able to run on such boards.

To solve this situation, we introduce support for subsystems to specify
their own custom system heap size requirement. Subsystems do
this by defining Kconfig options with the prefix HEAP_MEM_POOL_ADD_SIZE_.
The final value of the system heap is the sum of the custom
minimum requirements, or the value existing HEAP_MEM_POOL_SIZE option,
whichever is greater.

We also introduce a new HEAP_MEM_POOL_IGNORE_MIN Kconfig option which
applications can use to force a lower value than what subsystems have
specficied, however this behavior is disabled by default.

Whenever the minimum is greater than the requested value a CMake warning
will be issued in the build output.

This patch ends up modifying several places outside of kernel code,
since the presence of the system heap is no longer detected using a
non-zero CONFIG_HEAP_MEM_POOL_SIZE value, rather it's now detected using
a new K_HEAP_MEM_POOL_SIZE value that's evaluated at build.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
boards: arm: Use HEAP_MEM_POOL_ADD_SIZE KConfig options

Kconfig options with a HEAP_MEM_POOL_ADD_SIZE_ prefix should be used to
set the minimum required system heap size. This helps prevent
applications from creating a non-working image by trying to set a too
small value.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
boards: arm64: Use HEAP_MEM_POOL_ADD_SIZE KConfig options

Kconfig options with a HEAP_MEM_POOL_ADD_SIZE_ prefix should be used to
set the minimum required system heap size. This helps prevent
applications from creating a non-working image by trying to set a too
small value.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
boards: posix: Use HEAP_MEM_POOL_ADD_SIZE KConfig options

Kconfig options with a HEAP_MEM_POOL_ADD_SIZE_ prefix should be used to
set the minimum required system heap size. This helps prevent
applications from creating a non-working image by trying to set a too
small value.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
boards: riscv: Use HEAP_MEM_POOL_ADD_SIZE KConfig options

Kconfig options with a HEAP_MEM_POOL_ADD_SIZE_ prefix should be used to
set the minimum required system heap size. This helps prevent
applications from creating a non-working image by trying to set a too
small value.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
lib: acpi: Introduce HEAP_MEM_POOL_ADD_SIZE_ACPI option

All x86 boards have so far set a custom heap memory pool size because of
their dependency on ACPI. It makes more sense to introduce a new
ACPI-specific option, utilizing the recently added
HEAP_MEM_POOL_ADD_SIZE_ Kconfig option prefix, and adjust the default
value as necessary for each board.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
boards: xtensa: Use HEAP_MEM_POOL_ADD_SIZE KConfig options

Kconfig options with a HEAP_MEM_POOL_ADD_SIZE_ prefix should be used to
set the minimum required system heap size. This helps prevent
applications from creating a non-working image by trying to set a too
small value.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
soc: arm: Use HEAP_MEM_POOL_ADD_SIZE KConfig options

Kconfig options with a HEAP_MEM_POOL_ADD_SIZE_ prefix should be used to
set the minimum required system heap size. This helps prevent
applications from creating a non-working image by trying to set a too
small value.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
soc: xtensa: Use HEAP_MEM_POOL_ADD_SIZE KConfig options

Kconfig options with a HEAP_MEM_POOL_ADD_SIZE_ prefix should be used to
set the minimum required system heap size. This helps prevent
applications from creating a non-working image by trying to set a too
small value.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

tgorochowik commented on kernel/mempool.c:
```diff
@@ -56,9 +56,12 @@ void k_free(void *ptr)
 }
 
-#if (CONFIG_HEAP_MEM_POOL_SIZE > 0)
+#define HEAP_MEM_POOL_SIZE MAX(CONFIG_HEAP_MEM_POOL_SIZE, \
```
I appreciate the fix, we've also run into that issue recently.

Personally I think it would be better to just fail here if current size < min size, I know that that would mean we need to modify multiple samples, however with this approach this can be confusing to the users (i.e. they're getting a different size than they're setting).

jhedberg replied:
The question is, how exactly would we modify the applications? You could have them set the heap to the "worst case" scenario, but then that would result in something that doesn't build on boards with small amounts of RAM. You could do board-specific overlays, but that doesn't cover subsystems or SoCs that have specific requirements, and it'd also multiply the extent of changes (you'd need the same overlays for every single app that currently sets a heap size in their prj.conf).

I agree that it's not ideal that the rounding up happens silently. I was considering using something like `#warning HEAP_MEM_POOL_SIZE is less than HEAP_MEM_POOL_MIN_SIZE` but wouldn't that potentially also fail the build when `-Werror` is used?

finikorg replied:
I think cmake `message(WARNING ....)` should be enough

tgorochowik replied:
a warning sounds good

jhedberg replied:
Yep, I tried it and seems to work fine. I get this in the build output now:
```
CMake Warning at /Users/jhedberg/src/zephyr/zephyr/kernel/CMakeLists.txt:122 (message):


      CONFIG_HEAP_MEM_POOL_SIZE is less than CONFIG_HEAP_MEM_POOL_MIN_SIZE.
      The size of the system heap will be rounded up to the minimum size.
```

I've now pushed an update with this change.

nordicjm commented on kernel/Kconfig:
```diff
@@ -833,15 +833,29 @@ config KERNEL_MEM_POOL
 
 if KERNEL_MEM_POOL
 
+config HEAP_MEM_POOL_MIN_SIZE
+	int
+	default 1024 if POSIX_MQUEUE
+	default 0
```
This is a step in the right direction, but it's a fake solution, i.e. it gives the impression that it can be set to the minimum heap required for functionality but will be trumped by other Kconfigs.

Assume an application enables 3 features, one sets this to 3KB, another set it to 0, another sets it to 1MB - what is the actual value you get here? It's actually undefined, because it depends on the order in which the Kconfigs are encountered. Therefore with this huge issue, I cannot +1 this idea

jhedberg replied:
> This is a step in the right direction, but it's a fake solution, i.e. it gives the impression that it can be set to the minimum heap required for functionality but will be trumped by other Kconfigs.
> 
> Assume an application enables 3 features, one sets this to 3KB, another set it to 0, another sets it to 1MB - what is the actual value you get here? It's actually undefined, because it depends on the order in which the Kconfigs are encountered. Therefore with this huge issue, I cannot +1 this idea

Yes, that's correct. I was aware of this more fundamental issue when I filed #65722 and after the discussion there proceeded to create this PR. AFAIK there is no existing mechanism in the Zephyr build system currently to select the maximum from a set of multiple build-time defined values (whether it's Kconfig or something else). If there is, then that would definitely be a better solution to the problem.

The situation as it stands today isn't quite as bad as you describe though, since the defaults are not being set "all over the place" by "features", rather it's currently only boards or SoCs that set default values. So in that sense I do think this PR is taking the situation in a better direction. E.g. right now there are 11 tests applications under `tests/kernel` that all set explicit heap sizes that are too small for any ACPI-using board. The apps build fine, but result in runtime failures with misleading error messages on all of these boards.

nordicjm replied:
It's introducing a known defect to try and address an issue, so instead of the original 1 issue, you now have 2 issues. Agreed, this cannot be fixed in Kconfig as it stands today, I do not have a proposal for how this can be addressed, just that this is not the way to solve it

jhedberg replied:
Yep, basically what this does is to group things to "the application" and "everything else (kernel, subsystems, etc)". The second category is left on its own to figure out a consistent minimum value. That problem is not solved by this PR, although it kind of does work due the limited set of places where defaults are attempted to be set.

I see a few more potential complexities:
 - Should it really be `MAX(A, B, C, D)`? What if the submodules run together and at the same time need successful memory allocations? Wouldn't it then be `A + B + C + D` that we need?
 - It's not strictly that component A always has a specific requirement, rather the requirement might depend on what other components a component is combined with. E.g. ACPI always requires some substantial amount of heap, but the exact amount depends on what platform you use it on (it can range from a few hundred kilobytes to several megabytes).

jhedberg replied:
@nordicjm aggregating a list of different minimum system heap size requirements at build time, and then picking the highest value (also at build time), could be done using a cmake global property, i.e. each place (SoC, board, subsystem, etc) that has a specific minimum requirement would add something like the following to their CMakeLists.txt file:
```cmake
set_property(GLOBAL APPEND PROPERTY HEAP_MEM_POOL_REQ 32768)
```

This would replace the `HEAP_MEM_POOL_MIN_SIZE` Kconfig option I've so far been proposing. Does that sound like an acceptable solution for you?

jhedberg replied:
> ```cmake
> set_property(GLOBAL APPEND PROPERTY HEAP_MEM_POOL_REQ 32768)
> ```
> 

The above could also be made slightly nicer looking by using a helper function, assuming it's ok to add such a thing to `cmake/modules/extensions.cmake`. In that case all places needing to declare their requirement would just do something like:
```cmake
zephyr_heap_mem_pool_req(32768)
```

andyross replied:
Drive-by comment: I like the idea, but as someone who fantasizes about a cmake-free build future, I don't really love the idea of adding another dependency.  Also cmake is surprisingly opaque about "how to tell if this dependency is actually going to be built", IIRC.  It's not enough to do it by subsystem, you need to know if it was actually linked.

Where we've wanted to do this in the past (collecting settings/metadata from unrelated subsystems based on whether they're built or not) it's worked pretty well to rely on the linker.  Consider a special section containing heap requirements  that we could parse out of an intermediate link step the same way we do for SYS_INIT or IRQ_CONNECT declarations?

dleach02 replied:
Is there a way to make a Kconfig "additive"? For example, if feature X is enabled add <n> to a value? 

I've run across this type of thing with stack sizes where adding features should add to the stack size...

nordicjm replied:
> Is there a way to make a Kconfig "additive"?

No, Kconfig is so basic that you can't even have a Kconfig option that adds a number to another number (or option)

jhedberg replied:
@andyross thanks for the proposal! I'll explore that next.

Do people here have any thoughts on whether the end result should be max(a,b,c,d) or a+b+c+d? Should it perhaps be guided by a Kconfig option?

tejlmand replied:
another possibility would be to introduce a `max(a,b)` function as Kconfig extension command, like we have with `dt` functions.

That would not be a Kconfig symbol as we know it, but instead using variables, and therefore potentially more errorprone.

But posting it so we can discuss all options.
Here https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/kconfig/kconfigfunctions.py define a `max(a,b)`: 
```diff
diff --git a/scripts/kconfig/kconfigfunctions.py b/scripts/kconfig/kconfigfunctions.py
index 621ec3cb3a8..50bfcb5e520 100644
--- a/scripts/kconfig/kconfigfunctions.py
+++ b/scripts/kconfig/kconfigfunctions.py
@@ -756,6 +756,8 @@ def shields_list_contains(kconf, _, shield):
 
     return "y" if shield in list.split(";") else "n"
 
+def kconfigmax(kconf, _, a, b):
+    return max(a,b)
 
 # Keys in this dict are the function names as they appear
 # in Kconfig files. The values are tuples in this form:
@@ -807,4 +809,5 @@ functions = {
         "dt_nodelabel_array_prop_has_val": (dt_nodelabel_array_prop_has_val, 3, 3),
         "dt_gpio_hogs_enabled": (dt_gpio_hogs_enabled, 0, 0),
         "shields_list_contains": (shields_list_contains, 1, 1),
+        "max": (kconfigmax, 2, 2),
 }
```

allows us to do following in Kconfig files:
```
# Kconfig file A
HEAP_MEM_POOL_MIN_SIZE := $(max,$(HEAP_MEM_POOL_MIN_SIZE),2048)
```
and later
```
# Kconfig file B
HEAP_MEM_POOL_MIN_SIZE := $(max,$(HEAP_MEM_POOL_MIN_SIZE),8192)
```
or
```
# Kconfig file B
HEAP_MEM_POOL_MIN_SIZE := $(max,$(HEAP_MEM_POOL_MIN_SIZE),512)
```

and then in [kernel/Kconfig](https://github.com/zephyrproject-rtos/zephyr/pull/65908/files#diff-05292158cc01da0f208bda8b8a77b9bd2f2a4fdf6f259259d99f18b8688575fc) have:
```
config HEAP_MEM_POOL_SIZE
	int "Heap memory pool size (in bytes)"
	default $(HEAP_MEM_POOL_MIN_SIZE)
```

Now, the use of `HEAP_MEM_POOL_MIN_SIZE` as a variable requires that all `HEAP_MEM_POOL_MIN_SIZE := ...` assignment happens **before** `kernel/Kconfig` is sourced, but that is still far better than today where the first `default <value>` always win.

Combining this with an `add(a,b)` will even allow subsys to decide whether to do a max or add, as they seem fit.
So subsys A can do:
```
# A requires min 2048, but can live side-by-side with other modules.
HEAP_MEM_POOL_MIN_SIZE := $(max,$(HEAP_MEM_POOL_MIN_SIZE),2048)
```
but B can do:
```
# B requires additional 1024, regardless of current min.
HEAP_MEM_POOL_MIN_SIZE := $(add,$(HEAP_MEM_POOL_MIN_SIZE),1024)
```

tejlmand replied:
downsides is of course such approach is more error prone.
First, it's harder to debug as you cannot see who decided the value in case someone did this by mistake:
```
HEAP_MEM_POOL_MIN_SIZE := 128
```
The alternative of
```
config HEAP_MEM_POOL_MIN_SIZE_FOO
        int
        default 128
```
allows developers to easily see where HEAP_MEM_POOL_MIN_SIZE_FOO was defined.

Also mistakes like this can be hard to trace down:
```
HEAP_MEM_POOL_SIZE_MIN := $(max,$(HEAP_MEM_POOL_MIN_SIZE),4096)
```

tejlmand replied:
I lean mostly towards the combined Kconfig+CMake approach (https://github.com/zephyrproject-rtos/zephyr/pull/65908#discussion_r1418931869), but there is also something nice about staying purely in Kconfig.
Overall I would say 60/40 in favor of combined Kconfig+CMake approach, but would like to hear what others think.

jhedberg replied:
@tejlmand thanks for the suggestion! I think I'll go with the "Kconfig prefix" approach then.

To cater for subsystems wanting their requirement to be additive vs subsystems that can deal with sharing the heap with others, we could have two different prefixes, e.g. HEAP_MEM_POOL_MIN_SIZE_* and HEAP_MEM_POOL_ADD_SIZE_*. This would then result in a calculation max(a,b,c) + sum(x,y,z) where the first term comes from MIN_SIZE_* values and the second from ADD_SIZE_* values? Is this getting too complicated? 😄 I could in a first iteration only cover the max(a,b,c) case.

Commit:
posix: mqueue: Remove custom default for HEAP_MEM_POOL_SIZE

Use the new HEAP_MEM_POOL_ADD_SIZE_ prefix to construct a minimum
requirement for posix message queue usage. This way we can remove the
"special case" default values from the HEAP_MEM_POOL_SIZE Kconfig
definition.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
I've now pushed a major update which implements the proposal from @tejlmand of supporting Kconfig options with a `HEAP_MEM_POOL_MIN_SIZE_*` prefix that subsystems may define. I also added a `HEAP_MEM_POOL_IGNORE_MIN` option which can be enabled to let the application force a smaller value than the minimum, although this option is disabled by default (to allow existing in-tree tests to run correctly).

Another notable change is that now `kernel/CMakeLists.txt` evaluates what the heap size should be, and passes this to the build through a new `K_HEAP_MEM_POOL_SIZE` define. This new define should be used to determine whether the system heap is available, whereas until now the code was using `CONFIG_HEAP_MEM_POOL_SIZE > 0` to do this.

The Compliance check failure ("undefined Kconfig option") seems to be a false positive resulting from the CMake code that looks for a given Kconfig option prefix:
```
  import_kconfig(CONFIG_HEAP_MEM_POOL_MIN_SIZE_ ${DOTCONFIG} min_size_keys)
```

@nordicjm could you take another look?

Commit:
manifest: Update espressif revision to support system heap changes

This is needed so that module uses K_HEAP_MEM_POOL_SIZE for determining
the availablility of the k_malloc family of functions.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
manifest: Update libmetal revision to support system heap changes

This is needed so that module uses K_HEAP_MEM_POOL_SIZE for determining
the availablility of the k_malloc family of functions.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
Rebased to fix merge conflicts

jhedberg commented:
@nordicjm @tejlmand would be good to get some feedback on whether this is now an acceptable way forward. We could then proceed with merging the submodule PRs and then update the west.yml changes in this PR.

tejlmand commented on kernel/CMakeLists.txt:
```diff
@@ -117,6 +117,35 @@ target_sources_ifdef(CONFIG_OBJ_CORE              kernel PRIVATE obj_core.c)
+  set(max_min_size 0)
+
+  foreach(min_size ${min_size_keys})
+    if(${${min_size}} GREATER ${max_min_size})
```
as a general safety precaution, please quote the variables when expanding.
Empty values / values containing spaces may give very strange and hard to find issues.
```suggestion
  if("${${min_size}}" GREATER "${max_min_size}")
```

jhedberg replied:
Done

tejlmand commented on kernel/CMakeLists.txt:
```diff
@@ -117,6 +117,35 @@ target_sources_ifdef(CONFIG_OBJ_CORE              kernel PRIVATE obj_core.c)
+
+  foreach(min_size ${min_size_keys})
+    if(${${min_size}} GREATER ${max_min_size})
+      set(max_min_size ${${min_size}})
```
```suggestion
    set(max_min_size "${${min_size}}")
```

tejlmand commented on kernel/CMakeLists.txt:
```diff
@@ -117,6 +117,35 @@ target_sources_ifdef(CONFIG_OBJ_CORE              kernel PRIVATE obj_core.c)
+    endif()
+  endforeach()
+
+  if(CONFIG_HEAP_MEM_POOL_SIZE LESS max_min_size)
```
when testing, then it is allowed to use the variable name directly like this, but in the lines above you where de-referencing, like `GREATER ${max_min_size}`.

So for consistency, maybe dereference `max_min_size` here also.

jhedberg replied:
Done

tejlmand commented on kernel/CMakeLists.txt:
```diff
@@ -117,6 +117,35 @@ target_sources_ifdef(CONFIG_OBJ_CORE              kernel PRIVATE obj_core.c)
+    # Only warn if default value 0 has been modified
+    if(NOT CONFIG_HEAP_MEM_POOL_SIZE EQUAL 0)
+      message(WARNING "
+      CONFIG_HEAP_MEM_POOL_SIZE is less than HEAP_MEM_POOL_MIN_SIZE:
```
Good to warn the user, but I think would be even better if we also inform the user about which value we're going to use.

jhedberg replied:
> Good to warn the user, but I think would be even better if we also inform the user about which value we're going to use.

Sure, makes sense. I've now added this.

tejlmand commented on kernel/CMakeLists.txt:
```diff
@@ -117,6 +117,35 @@ target_sources_ifdef(CONFIG_OBJ_CORE              kernel PRIVATE obj_core.c)
+
+  zephyr_compile_definitions(K_HEAP_MEM_POOL_SIZE=${max_min_size})
+else()
+  zephyr_compile_definitions(K_HEAP_MEM_POOL_SIZE=0)
```
This else is taken when `CONFIG_HEAP_MEM_POOL_SIZE==0`.

Basically this requires that a lib / subsys is which does `HEAP_MEM_POOL_SIZE_FOO=<value>` is also required to ensure `CONFIG_HEAP_MEM_POOL_SIZE!=0`.

Is that intentionally ?

jhedberg replied:
This is the else-branch for `if(${CONFIG_KERNEL_MEM_POOL})`, in which case the entire `kernel/mempool.c` will not be included in the build.

jhedberg replied:
> This is the else-branch for `if(${CONFIG_KERNEL_MEM_POOL})`, in which case the entire `kernel/mempool.c` will not be included in the build.

That said, maybe we want a warning if any non-zero `HEAP_MEM_POOL_MIN_SIZE_FOO` options exist but the user has disabled `KERNEL_MEM_POOL` (it defaults to enabled).

jhedberg replied:
> > This is the else-branch for `if(${CONFIG_KERNEL_MEM_POOL})`, in which case the entire `kernel/mempool.c` will not be included in the build.
> 
> That said, maybe we want a warning if any non-zero `HEAP_MEM_POOL_MIN_SIZE_FOO` options exist but the user has disabled `KERNEL_MEM_POOL` (it defaults to enabled).

Actually, not quite sure about this. The current idea is that `kernel/CMakeLists.txt` doesn't even both looking for and parsing MIN_SIZE options if `KERNEL_MEM_POOL` is not set (the idea was to avoid the build time overhead if it serves no useful purpose)

jhedberg replied:
And I just realised that `CONFIG_HEAP_MEM_POOL_SIZE` till be unset if `CONFIG_KERNEL_MEM_POOL` is unset (since the latter depends on the former), so we should probably follow the same for `K_HEAP_MEM_POOL_SIZE`, i.e. not define it at all in that case.

tejlmand reviewed (changes requested):
I think this is the most sane approach to achieve what is desired when considering the limitations of Kconfig.

Some minor adjustments needed.

tejlmand commented:
> The Compliance check failure ("undefined Kconfig option") seems to be a false positive resulting from the CMake code that looks for a given Kconfig option prefix:

You need to add `HEAP_MEM_POOL_MIN_SIZE_ ` to this list:
https://github.com/zephyrproject-rtos/zephyr/blob/2f230e89036d9595b1719118a7695db488f3ba14/scripts/ci/check_compliance.py#L614

Commit:
scripts: compliance: Add HEAP_MEM_POOL_ADD_SIZE_ to undef whitelist

The HEAP_MEM_POOL_ADD_SIZE_ is used as a prefix for matching specific
Kconfig option names, i.e. it's not a real option in itself.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
@tejlmand Thanks for the review! I've now pushed an update which should (hopefully) take care of the issues you pointed out.

tejlmand commented on kernel/CMakeLists.txt:
```diff
@@ -117,8 +117,35 @@ target_sources_ifdef(CONFIG_OBJ_CORE              kernel PRIVATE obj_core.c)
+      set(final_heap_size "${max_min_size}")
+    endif()
+    # Only warn if default value 0 has been modified
+    if(NOT CONFIG_HEAP_MEM_POOL_SIZE EQUAL 0)
```
CONFIG_HEAP_MEM_POOL_SIZE can never be zero because this line evaluates to false when `CONFIG_HEAP_MEM_POOL_SIZE =0`:
https://github.com/zephyrproject-rtos/zephyr/blob/4a199d7f0fd43b5d18e21eec2f1651fad52e853c/kernel/CMakeLists.txt#L118

tejlmand replied:
scratch that, somehow I was reading this as: `if(${CONFIG_HEAP_MEM_POOL_SIZE})`

tejlmand commented on kernel/Kconfig:
```diff
@@ -841,7 +841,18 @@ config HEAP_MEM_POOL_SIZE
 	  dynamically allocating memory using k_malloc(). The maximum size of
 	  the memory pool is only limited to available memory. A size of zero
-	  means that no heap memory pool is defined.
+	  means that no heap memory pool is defined. Note that if subsystems
+	  specify HEAP_MEM_POOL_MIN_SIZE_* values that are larger than the
+	  value given here, then the heap size will be rounded up to that value
```
> A size of zero means that no heap memory pool is defined. Note that if subsystems
> specify HEAP_MEM_POOL_MIN_SIZE_* values that are larger than the
> value given here, then the heap size will be rounded up to that value.

This must be made clearer.
From the current description it's  not fully clear if having `HEAP_MEM_POOL_SIZE=0` and `HEAP_MEM_POOL_MIN_SIZE_<name>=<value>` means that no heap mem pool will be defined or if it means that heap size will be rounded up to `<value>`.

jhedberg replied:
Thanks. I've now updated the Kconfig help text.

tejlmand reviewed (changes requested):
thanks for the update.

One thing not fully clear to me is whether it's intentionally that a subsys / lib setting a `HEAP_MEM_POOL_MIN_SIZE` must also set a `HEAP_MEM_POOL_SIZE` different from zero in order to work.

Notice that old code changed here: https://github.com/zephyrproject-rtos/zephyr/pull/65908/commits/1ef3d96d8a4c8f06b74b2b0db4d6e06f314d3f1e

has no `depends on KERNEL_MEM_POOL` meaning this code will **always** set a value for `KERNEL_MEM_POOL` which again means that `kernel/mempool.c` will be compiled and linked.

However, that's not the case anymore in 1ef3d96d8a4c8f06b74b2b0db4d6e06f314d3f1e cause the value of HEAP_MEM_POOL_SIZE may be zero, and thus the value of `HEAP_MEM_POOL_MIN_SIZE_<name>`  is ignored.

cfriedt commented:
+ @keith-packard, since he's done *a lot* of work in this area.

carlescufi reviewed (changes requested):
Please add a migration guide entry

carlescufi commented:
Architecture WG:

- @jhedberg explains the concept
- @henrikbrixandersen argues that we may want to follow the approach described [here](https://github.com/zephyrproject-rtos/zephyr/pull/65908#discussion_r1420190340) from the get-go
- @nashif argues that it would be simpler to just let each board allocate the whole of the remaining free memory instead
- @nordicjm replies that this is actually solving a problem: ACPI not being able to run because of small heap size
- @henrikbrixandersen suggests we replace the _MIN with _ADD instead of having them side by side

rodrigopex commented:
@jhedberg, great idea. Congrats on that. That is missing an exciting feature of the system. A suggestion about the naming. I don't know if there is still time for this. However, I would suggest using instead of prefixing with `HEAP_MEM_POOL_MIN_SIZE_`, we should use that as a suffix like `_HEAP_MEM_POOL_MIN_SIZE`. For example:

```c
CONFIG_HEAP_MEM_POOL_MIN_SIZE_ZBUS=512
// vs
CONFIG_ZBUS_HEAP_MEM_POOL_MIN_SIZE=512
```

To me, the second feels more natural than the first one. What do you think?

keith-packard commented:
Right now, malloc uses all of the unallocated memory in most configurations. That took a bit of fixing as it relies on `_end` being correctly defined, which was missing in a couple of cases. That's certainly a far easier than needing to manually tweak configuration values when subsystems change their memory usage. However, it also means that you don't know if there's enough memory for any particular task as there's no need to go and measure what a subsystem will use.

What I did for bare-metal picolibc applications was to let applications specify the minimum size of this heap; at link time, it checks to make sure there's at least that much unallocated memory, but then lets the application use as much as is available. If that were adjusted to provide a test mode where allocations beyond that amount would fail, then developers would verify that the application ran within its specified bounds but could exceed those in production.

The other thing that would be really good is if malloc could share this same heap where possible. I think that's any target without an MPU/MMU? Splitting a scarce resource like the heap at build time doesn't seem necessary to me.

jhedberg commented:
> @jhedberg, great idea. Congrats on that. That is a missing feature of the system.

Thanks! :)

> A suggestion about the naming. I don't know if there is still time for this. However, I would suggest using instead of prefixing with `HEAP_MEM_POOL_MIN_SIZE_`, we should use that as a suffix like `_HEAP_MEM_POOL_MIN_SIZE`. For example:
> 
> ```c
> CONFIG_HEAP_MEM_POOL_MIN_SIZE_ZBUS=512
> // vs
> CONFIG_ZBUS_HEAP_MEM_POOL_MIN_SIZE=512
> ```
> 
> To me, the second feels more natural than the first one. What do you think?

In some ways, yes, but there are a few reasons I wouldn't go for it:

 - There's no current API to collect/import options based on suffixes, just prefixes (`import_kconfig()`)
 - If you think about it, owning a namespace means you get to freely create stuff under it and define the meaning for the things under it (Kconfig options in this case). However here, the meaning (i.e. how the values get used) is defined by the system memory pool implementation (owned by the kernel). So in that sense having a common, kernel-owned namespace seems like the right choice.

jhedberg commented:
@keith-packard thanks for the insights! It's indeed suboptimal that we have two separate general purpose heaps right now. I think I'd still want to proceed with this PR, since it doesn't take away from the option of combining the system and libc heaps in the future. In some ways it kind of even paves the way toward it, since we now get a more accurate "minimum requirement" indication for the system heap.

keith-packard commented:
> @keith-packard thanks for the insights! It's indeed suboptimal that we have two separate general purpose heaps right now. I think I'd still want to proceed with this PR, since it doesn't take away from the option of combining the system and libc heaps in the future. In some ways it kind of even paves the way toward it, since we now get a more accurate "minimum requirement" indication for the system heap.

Yup, totally agree -- just letting you know that we'll have to fix the malloc pool to not also use all available memory at the same time as that won't go well. I think that means we should change the malloc heap to always use a fixed size heap before we have the system heap use all available memory. After that, we can look at merging the two heaps where possible.

Commit:
doc: migration-guide-3.6: Add a note about system heap size changes

Add a note about the new K_MEM_POOL_HEAP_SIZE define and the mechanism
for specifying custom system heap size requirements.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
As agreed in the Arch meeting, I've now pushed an update to use "ADD" instead of "MIN". I also added a new commit to document the changes in the migration guide. @nordicjm @tejlmand @carlescufi please have another look.

nordicjm commented on doc/kernel/memory_management/heap.rst:
```diff
@@ -165,6 +165,16 @@ the kernel not to define the heap memory pool object. The maximum size is limite
 
+In addition, each subsystem (board, driver, library, etc) can set a custom
+requirement by defining a Kconfig option with the prefix
+HEAP_MEM_POOL_ADD_SIZE. If multiple subsystems specify custom values, the
```
```suggestion
``HEAP_MEM_POOL_ADD_SIZE_`` (this value is in bytes). If multiple subsystems specify custom values, the
```

jhedberg replied:
Fixed

nordicjm commented on doc/kernel/memory_management/heap.rst:
```diff
@@ -165,6 +165,16 @@ the kernel not to define the heap memory pool object. The maximum size is limite
+tries to set a value that's less than the minimum value, this will be ignored
+and the minimum value will be used instead. To force a smaller than minimum
+value to be used, the application may enable the
+:kconfig:option:`CONFIG_HEAP_MEM_POOL_IGNORE_MIN` option, however this is
+discouraged as it may result in a non-working application image.
```
Not sure I would use these words, because the value **is** likely to be wrong, it should be the worst case scenario, as discussed yesterday people will need to optimise their end applications manually, would be good to capture that knowledge here.

jhedberg replied:
I rephrased this.

nordicjm commented on kernel/CMakeLists.txt:
```diff
@@ -117,8 +117,33 @@ target_sources_ifdef(CONFIG_OBJ_CORE              kernel PRIVATE obj_core.c)
+    endif()
+    # Only warn if default value 0 has been modified
+    if(NOT CONFIG_HEAP_MEM_POOL_SIZE EQUAL 0)
+      message(WARNING "
+      CONFIG_HEAP_MEM_POOL_SIZE is less than requested minimum:
+        ${CONFIG_HEAP_MEM_POOL_SIZE} < ${add_size_sum}
+      Setting the system heap size to ${final_heap_size}")
```
Can see this being useful for testing sample apps maybe but seems like it would get annoying pretty fast for an end application - thoughts @henrikbrixandersen ?

tejlmand replied:
I guess that's why @jhedberg created the `HEAP_MEM_POOL_IGNORE_MIN` setting.

https://github.com/zephyrproject-rtos/zephyr/blob/96ed1db313dfda872bf1b8637b8338e2e8578130/kernel/Kconfig#L853-L859

tejlmand replied:
actually it seems this warning may be printed even if `CONFIG_HEAP_MEM_POOL_IGNORE_MIN=y` because CONFIG_HEAP_MEM_POOL_SIZE is being compared against the calculated `add_size_sum` regardless of the ignore setting.

jhedberg replied:
Now updated to only print the warning if `IGNORE_MIN` was not set.

nordicjm commented on kernel/Kconfig:
```diff
@@ -840,8 +840,24 @@ config HEAP_MEM_POOL_SIZE
+	help
+	  This option can be set to force setting a smaller heap memory pool
+	  size than what's specified by enabled subsystems. There is a risk
+	  that this could create a non-functioning application, so this option
+	  should be used with great care.
```
Would probably want a note here also

jhedberg replied:
Done.

nordicjm commented on doc/releases/migration-guide-3.6.rst:
```diff
@@ -29,6 +29,15 @@ Build System
+* The system heap size and its availability is now determined by a ``K_MEM_POOL_HEAP_SIZE``
+  define instead of the :kconfig:option:`CONFIG_MEM_POOL_HEAP_SIZE` Kconfig option. Subsystems
+  can specify their own custom system heap size requirements by specifying Kconfig options with
+  the prefix ``CONFIG_MEM_POOL_HEAP_ADD_SIZE``. The old Kconfig option still exists, but will be
```
first commit says HEAP_MEM_POOL_ADD_SIZE, this says MEM_POOL_HEAP_ADD_SIZE - both can't be right? Plus should have _ suffix

jhedberg replied:
> first commit says HEAP_MEM_POOL_ADD_SIZE, this says MEM_POOL_HEAP_ADD_SIZE - both can't be right? Plus should have _ suffix

Good catch. Now fixed.

jhedberg commented:
Note that while the https://github.com/zephyrproject-rtos/hal_espressif/pull/256 module update has been merged (and west.yml updated accordingly in this PR), we still need to wait for https://github.com/zephyrproject-rtos/libmetal/pull/23 which in turn is blocked by #65523.

tejlmand commented on boards/x86/intel_adl/Kconfig.defconfig:
```diff
@@ -33,7 +33,7 @@ config ACPI
 
 if ACPI
-config HEAP_MEM_POOL_SIZE
+config HEAP_MEM_POOL_ADD_SIZE_ACPI
```
if various subsystems starts to use `ADD_SIZE` then the default for this board should probably be recalculated as the current default is probably chosen based on the subsystems / libs default enabled for this board.

But it would be the board maintainers who would actually know how to determine the optimal value for the `ADD_SIZE`.

This comment applies for boards in general where `HEAP_MEM_POOL_SIZE` is now changed to an `ADD_SIZE`.

I'm fine with the current approach of changing to `ADD_SIZE`, and then leave any optimization (heap size reduction) as followup work for board maintainers.

tejlmand reviewed (dismissed):
lgtm.

I think some followup work / task should be created to look into correct `ADD_SIZE` values because I believe a lot of current `HEAP_MEM_POOL_SIZE` are already taking into account some subsystems being enabled. Subsystems which may now define their own `ADD_SIZE`.

nordicjm commented on lib/acpi/Kconfig:
```diff
@@ -30,6 +30,10 @@ config ACPI_MAX_PRT_ENTRY
 endif # PCIE_PRT
 
+# Default minimum system heap required by ACPI
+config HEAP_MEM_POOL_ADD_SIZE_ACPI
+	def_int 1048576 # 1MB
```
Nothing to do here but in future we could have a template for this similar to the logging one

nordicjm commented:
@rodrigopex please check if `doc/services/zbus/index.rst` needs updating:
```
.. warning::
   ZBus uses :zephyr_file:`include/zephyr/net/buf.h` (network buffers) to exchange data with message
   subscribers. So, chose carefully the configurations
   :kconfig:option:`CONFIG_ZBUS_MSG_SUBSCRIBER_NET_BUF_POOL_SIZE` and
   :kconfig:option:`CONFIG_HEAP_MEM_POOL_SIZE`. They are crucial to a proper VDED execution
   (delivery garantee) considering message subscribers.
```

@firscity please check if `drivers/xen/Kconfig` needs updating:
```
config XEN_GRANT_TABLE
        bool "Xen grant table driver"
        depends on HEAP_MEM_POOL_SIZE > 0
        default y
        help
          Xen grant table driver. Please note that driver uses dynamic memory
          allocation with k_malloc(), so CONFIG_HEAP_MEM_POOL_SIZE should be
          >= number of pages, that you want to alloc and grant or foreign frames
          that you want to map.
```

@jukkar @rlubos please check if `include/zephyr/net/buf.h` needs updating:
```
/**
 *
 * @brief Define a new pool for buffers using the heap for the data.
 *
 * Defines a net_buf_pool struct and the necessary memory storage (array of
 * structs) for the needed amount of buffers. After this, the buffers can be
 * accessed from the pool through net_buf_alloc. The pool is defined as a
 * static variable, so if it needs to be exported outside the current module
 * this needs to happen with the help of a separate pointer rather than an
 * extern declaration.
 *
 * The data payload of the buffers will be allocated from the heap using
 * k_malloc, so CONFIG_HEAP_MEM_POOL_SIZE must be set to a positive value.
 * This kind of pool does not support blocking on the data allocation, so
 * the timeout passed to net_buf_alloc will be always treated as K_NO_WAIT
 * when trying to allocate the data. This means that allocation failures,
 * i.e. NULL returns, must always be handled cleanly.
```

@cfriedt please check if `lib/posix/mqueue.c` needs updating:
```
/**
 * @brief Open a message queue.
 *
 * Number of message queue and descriptor to message queue are limited by
 * heap size. increase the size through CONFIG_HEAP_MEM_POOL_SIZE.
 *
 * See IEEE 1003.1
 */
```

jhedberg commented:
Rebased, and updated the libmetal reference to point at https://github.com/zephyrproject-rtos/libmetal/pull/25 instead of https://github.com/zephyrproject-rtos/libmetal/pull/23

jhedberg commented:
..and now rebased again with the final libmetal SHA after https://github.com/zephyrproject-rtos/libmetal/pull/25 was merged

Now all CI checks should (hopefully) be green.

@nordicjm thanks for the thorough analysis of places that most likely still require updating. I think those updates can be done gradually after this PR has been merged. The idea with having similar template support as we have for logging kconfig symbols sounds like a nice follow-up improvement as well.

tejlmand commented on boards/posix/nrf_bsim/Kconfig.defconfig:
```diff
@@ -44,7 +44,8 @@ config BT_CTLR
 	depends on BT
 
-config HEAP_MEM_POOL_SIZE
+config HEAP_MEM_POOL_ADD_SIZE_BOARD
+	int
```
generally it's bad practice to define Kconfig settings in defconfig, the Kconfig.defconfig should only adjust existing defaults (which was the case when using `HEAP_MEM_POOL_SIZE`.

Comment applies to other updated Kconfig.defconfig as well.

jhedberg replied:
Sure, that's a good point. I'll do it in a follow-up PR.

tejlmand reviewed (approved):
Approved.

Minor nit observed. Kconfig.defconfigs should generally not define settings, only adjust defaults of settings defined elsewhere.
But as this PR is changing HEAP_MEM_POOL_SIZE -> HEAP_MEM_POOL_ADD_SIZE, then moving this to the board's Kconfig can done as followup.

finikorg commented on boards/x86/qemu_x86/Kconfig.defconfig:
```diff
@@ -53,10 +49,6 @@ config BOARD
-config HEAP_MEM_POOL_SIZE
-	default 1048576 if ACPI
-	depends on KERNEL_MEM_POOL
-
```
So qemu does not use ACPI anymore?

jhedberg replied:
> So qemu does not use ACPI anymore?

No it does. The Kconfig option for ACPI comes from `lib/acpi/Kconfig` and since its default value is 1M (which is what qemu used) we don't need to modify this default in the qemu defconfig anymore.

carlescufi reviewed (approved):
Looks great now, thank you!

nordic-krch commented:
I know that it's merged already but it would be better if `<option>_ADD_<subsys>` pattern were followed (`HEAP_MEM_POOL_SIZE_ADD_` instead of `HEAP_MEM_POOL_ADD_SIZE_`). I guess that this would allow to wrap cmake operation into a function. As you can see in #67854 same approach is taken for channel count where same cmake code need to be copied and modified.
